### PR TITLE
Creates a script to back populate the level_source_id column of the user_levels table.

### PR DESCRIPTION
**Pull request by ashercodeorg**
After establishing safe values for batch_size and delay_sec, I'll change this summary to reflect them. At the moment, I have no idea what values will be powerful enough to make progress and weak enough to not cripple production.

### CODE REVIEW

File: bin/oneoff/back_populate_user_levels_level_source_id [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+batch_size = ARGV[2].to_i
+delay_sec = ARGV[3].to_f
+
+until start_index > end_index
+  UserLevel.
+    where(id: start_index..(start_index + batch_size)).
+    where(level_source_id: nil).each do |user_level|
+
+    activity = Activity.
+      where(user_id: user_level.user_id, level_id: user_level.level_id).last
```
Comment (joshlory) [inline bin/oneoff/back_populate_user_levels_level_source_id:31]: Do we also need to filter by script?

Comment (ashercodeorg) [inline bin/oneoff/back_populate_user_levels_level_source_id:31]: For better or for worse, the `activities` table does not have `script_id` information.

```
DESCRIBE activities;
+-----------------+--------------+------+-----+---------+----------------+
| Field           | Type         | Null | Key | Default | Extra          |
+-----------------+--------------+------+-----+---------+----------------+
| id              | int(11)      | NO   | PRI | NULL    | auto_increment |
| user_id         | int(11)      | YES  | MUL | NULL    |                |
| level_id        | int(11)      | YES  |     | NULL    |                |
| action          | varchar(255) | YES  |     | NULL    |                |
| url             | varchar(255) | YES  |     | NULL    |                |
| created_at      | datetime     | YES  |     | NULL    |                |
| updated_at      | datetime     | YES  |     | NULL    |                |
| attempt         | int(11)      | YES  |     | NULL    |                |
| time            | int(11)      | YES  |     | NULL    |                |
| test_result     | int(11)      | YES  |     | NULL    |                |
| level_source_id | int(11)      | YES  | MUL | NULL    |                |
| lines           | int(11)      | NO   |     | 0       |                |
+-----------------+--------------+------+-----+---------+----------------+
```

Comment (joshlory) [inline bin/oneoff/back_populate_user_levels_level_source_id:31]: Ah right, that's is the weird quirk of `last_attempt` vs. dot color.  A level can load a last attempt from a different script even though the dot (`UserLevel`) is 'not_tried'.

**Comment by Hamms** [comment]
Might want to put a comment in somewhere noting that `end_index` is not strictly enforced if `end_index - start_index` is not a multiple of `batch_size`

edit: or enforce it

File: bin/oneoff/back_populate_user_levels_level_source_id [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+    user_level.update(level_source_id: activity[:level_source_id])
```
Comment (Hamms) [inline bin/oneoff/back_populate_user_levels_level_source_id]: should we consider skipping validation?

Comment (ashercodeorg) [inline bin/oneoff/back_populate_user_levels_level_source_id]: Yes. And I (already) rejected the consideration. Given that (a) I intend to babysit the script while it runs, (b) the script is idempotent, (c) the script does no harm if it bombs out, I'd just assume throw and see the exception if validation fails.

**Comment by ashercodeorg** [comment]
Given the intent of the script, the idempotence of the script, that it is a oneoff script, and the lack of harm caused by processing a few "extra" records, I'm happy to say "User Beware" for the edge case when `end_index - start_index % batch_size != 0`.

File: bin/oneoff/back_populate_user_levels_level_source_id (hunk 2 above)
Comment (joshlory) [inline bin/oneoff/back_populate_user_levels_level_source_id]: In that case I think you want `.update!`?

Comment (ashercodeorg) [inline bin/oneoff/back_populate_user_levels_level_source_id]: Oops, yes. Done.

**Comment by joshlory** [comment]
FWIW `last_attempt` includes `.order('id desc')`.  I'm not sure if this is necessary?

``` ruby
def last_attempt(level)
  Activity.where(user_id: self.id, level_id: level.id).order('id desc').first
end
```

File: bin/oneoff/back_populate_user_levels_level_source_id [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+until start_index > end_index
+  UserLevel.
+    where(id: start_index..(start_index + batch_size)).
+    where(level_source_id: nil).each do |user_level|
+
+    activity = Activity.
+      where(user_id: user_level.user_id, level_id: user_level.level_id).last
+
+    user_level.update!(level_source_id: activity[:level_source_id])
```
Comment (joshlory) [inline bin/oneoff/back_populate_user_levels_level_source_id]: You may want to defend against nil `activity`.  Could also skip the update if `:level_source_id` is nil (can this ever happen?).

**Comment by joshlory** [comment]
Added two more questions, otherwise LGTM.  Thank you for doing this!

**Comment by ashercodeorg** [comment]
Hmm. My belief is that `Activity.where(user_id: self.id, level_id: level.id).order('id desc').first` can be written more succinctly (possibly also with a more efficient execution) as `Activity.where(user_id: self.id, level_id: level.id).last` given that `Activity.id` is the primary key of `Activity`.

Reference: http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-last

File: bin/oneoff/back_populate_user_levels_level_source_id (hunk 3 above)
Comment (ashercodeorg) [inline bin/oneoff/back_populate_user_levels_level_source_id]: For now, I'm fine with throwing (an uncaught) `NoMethodError` exception if `activity` is nil.

Going to change things to throw an exception if `level_source_id` is nil.

**Pull request merged**